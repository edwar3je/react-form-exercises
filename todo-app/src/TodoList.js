import './TodoList.css';
import { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';

const TodoList = () => {
    
    const initialState = [{
        id: uuidv4(),
        task: "Water the plants."
    }];

    const [todos, setTodos] = useState(initialState);
    
    const addTodo = (task, id) => {
        setTodos(todos => [...todos, { task, id }])
    };

    const removeTodo = (uniqueId) => {
        setTodos(
            todos => todos.filter(
                ({ id }) => id !== uniqueId
            )
        )
    };

    /*function removeTodo(uniqueId) {
        return setTodos(
            todos => todos.filter(
                ({ id }) => id !== uniqueId
            )
        )
    };*/

    return (
        <div>
            <NewTodoForm add={addTodo}/>
            <div className="todos-container">
                {todos.map(({task, id}) => <Todo task={task} remove={removeTodo} id={id} key={id}/>)}
            </div>
        </div>
    );
}

export default TodoList;