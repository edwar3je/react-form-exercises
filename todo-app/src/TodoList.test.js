import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

it("renders without crashing", () => {
    render(<TodoList />);
});

it("matches the snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});

it("displays initial data from useState", () => {
    const { queryByText } = render(<TodoList />);
    const firstTask = queryByText("Water the plants.");
    expect(firstTask).toBeInTheDocument();
});

it("adds a task to the DOM", () => {
    // expect the task we want to add to not be in the document before form submission
    const { queryByText, queryByLabelText } = render(<TodoList />);
    expect(queryByText("Make the bed.")).not.toBeInTheDocument();

    // extract form DOM elements and use fireChange to submit form
    const taskInput = queryByLabelText("Task");
    const submitButton = queryByText("Add New Task");
    fireEvent.change(taskInput, {target: {value: "Make the bed."}});
    fireEvent.click(submitButton);

    // expect the newly added task to be in the document
    const secondTask = queryByText("Make the bed.");
    expect(secondTask).toBeInTheDocument();
});

it("makes input value a blank string after adding a task to the DOM", () => {
    const { queryByText, queryByLabelText } = render(<TodoList />);

    // extract form DOM elements and use fireChange to submit form
    const taskInput = queryByLabelText("Task");
    const submitButton = queryByText("Add New Task");
    fireEvent.change(taskInput, {target: {value: "Make the bed."}});
    fireEvent.click(submitButton);

    // demonstrate that the form input's value is now blank
    expect(taskInput.value).toBe("");
});

it("removes a task from the DOM", () => {
    // expect the task we want to add to not be in the document before form submission
    const { queryByText, queryByLabelText, container } = render(<TodoList />);
    const firstTask = queryByText("Water the plants.")
    const firstRemoveButton = container.querySelector("button#Wat")
    expect(firstTask).toBeInTheDocument();
    expect(firstRemoveButton).toBeInTheDocument();

    // extract form DOM elements and use fireChange to submit form
    const taskInput = queryByLabelText("Task");
    const submitButton = queryByText("Add New Task");
    fireEvent.change(taskInput, {target: {value: "Make the bed."}});
    fireEvent.click(submitButton);

    // expect the newly added task to be in the document
    const secondTask = queryByText("Make the bed.");
    const secondRemoveButton = container.querySelector("button#Mak");
    expect(secondTask).toBeInTheDocument();
    expect(secondRemoveButton).toBeInTheDocument();

    // demonstrate that the first task was removed (and the second task wasn't)
    fireEvent.click(firstRemoveButton);
    expect(firstTask).not.toBeInTheDocument();
    expect(secondTask).toBeInTheDocument();

    // demonstrate that the second task was removed
    fireEvent.click(secondRemoveButton);
    expect(secondTask).not.toBeInTheDocument();
});