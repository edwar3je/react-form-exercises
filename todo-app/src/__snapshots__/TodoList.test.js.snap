// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="todo-form"
    >
      <label
        for="task"
      >
        Task
      </label>
      <input
        class="task"
        id="task"
        placeholder="todo"
        type="task"
      />
      <button
        class="submit-button"
        type="submit"
      >
        Add New Task
      </button>
    </form>
    <div
      class="todos-container"
    >
      <div
        class="todo-container"
      >
        <p>
          Water the plants.
        </p>
        <button
          class="remove-button"
          id="Wat"
        >
          X
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
